ibis/tests/benchmarks/test_benchmarks.py

```diff
@@ -753,3 +753,80 @@ def parse_many(types):
 
     types = ["VARCHAR", "INTEGER", "DOUBLE", "BIGINT"] * 1000
     benchmark(parse_many, types)
+
+
+def generate_asof_join_benchmark_data(
+    nprices: int, ntransactions: int
+) -> tuple[pd.DataFrame, pd.DataFrame]:
+    # Generate synthetic time data for prices
+    prices_times = pd.date_range(start="09:00", periods=nprices, freq="T")
+
+    # Define potential transaction times (every 00 or 30 seconds)
+    transaction_times = pd.date_range(
+        start="09:00", periods=ntransactions * 2, freq="30S"
+    )
+
+    # Sample from the potential transaction times for transactions
+    times_transactions = np.random.choice(
+        transaction_times, size=ntransactions, replace=False
+    )
+
+    # Generate synthetic price data
+    prices = (
+        102.5 + np.sin(np.linspace(0, nprices, nprices) / 100) * 2.0
+    )  # Sin wave around 102.5 with amplitude of 2
+
+    # Generate synthetic transaction data
+    shares = np.random.randint(5, 20, size=ntransactions)
+
+    # Create DataFrames
+    prices_df = pd.DataFrame({"time": prices_times, "price": prices})
+    transactions_df = pd.DataFrame({"time": times_transactions, "shares": shares})
+
+    # Sort transactions_df by time
+    return prices_df, transactions_df.sort_values(by="time")
+
+
+@pytest.fixture
+def asof_sql() -> str:
+    return """
+    SELECT t.time, t.shares, p.price
+    FROM transactions t
+    ASOF JOIN prices p
```

Comment:
Duckdb's ASOF join is known to be slower than pandas', I'd expect that at some point there will be a major perf improvement here just due to duckdb further optimizing this op.

Any complicated-enough query can reproduce the performance difference between `.sql()` and `.execute()` though, here's a much simpler one using their already pretty zippy standard joins. Perhaps it would be better to avoid asof joins in this case to keep the benchmark focused more on the specific issue?

```python
from contextlib import contextmanager
import timeit

import pandas as pd
import numpy as np
import duckdb

import ibis


@contextmanager
def timer(name):
    start = timeit.default_timer()
    yield
    end = timeit.default_timer()
    elapsed = end - start

    print(f"{name}: {elapsed:.2f} seconds")


N = 20_000_000
id1 = np.arange(N)
np.random.shuffle(id1)
id2 = np.arange(N)
np.random.shuffle(id2)

t1 = pd.DataFrame({"id": id1, "x": np.arange(N)})
t2 = pd.DataFrame({"id": id2, "y": np.arange(N)})

con = duckdb.connect()
con.execute("CREATE TABLE t1 AS SELECT * FROM t1")
con.execute("CREATE TABLE t2 AS SELECT * FROM t2")

sql = "SELECT * FROM t1 INNER JOIN t2 ON t1.id = t2.id"

with timer("Using .sql"):
    con.sql(sql).arrow()

with timer("Using .execute"):
    con.execute(sql).arrow()
```

Reply:
Thanks, that's a good idea. Using `INNER JOIN` should be a bit more robust.

Reply:
Done!